content/function_calling.qmd

```diff
@@ -4,9 +4,146 @@ toc-title: '![](../imgs/cover.jpg){width=240px}<br> <h3>Function Calling</h3>'
+
+The planning module then decides which tools to use, if any, to answer the user request. The output of the planning module is a response message containing one or several tool calls (or a final answer). The agent then executes the tool calls by first parsing the response, then executing the functions. Based on the tool outputs, a final answer is generated and sent back to the user.
+
+
+### React agents
+
+There a several types of agent. Now, we want to fucus on the ReAct agent introduced by [@yaoReActSynergizingReasoning2023]. The ReAct agent is a type of agent that uses the ReAct framework to solve complex tasks by reasoning in multiple steps. It is based on the idea of "thought-action-observation" loops. The LLM is given a task and it generates a thought, which is then used to decide on an action. The action is executed and the observation is fed back into the LLM. This process is repeated until the LLM decides that it has enough information to answer the question or if the maximum number of iterations is reached.
+
+
+### Llamaindex
+
+LLamaindex is a framework that  makes it easy to implement and use agents. In Llamaindex an agent consists a an *agent runner* and an *agent_worker*. Think of the agent runner as the agent core in the architecture schematics and the agent worker as the planning module. The tools are functions implemented in python that can be executed by the agent worker. Finally, the memory module consists of a simple text buffer, logging the conversation history between the user and the agent and between the agent and the tools.
+
+
+::: {.callout-note}
+## 📝 Task
+
+Let's have a look!
+
+1. Open a notebook and connect it with a local LLM using LM Studio.
+2. Define a function that can be called by the agent to get the current weather in a given location. (Implement it this time, it doesn't need to work, just return random weather)
+3. Initialize a ReAct agent using LLamaindex (you can use [this tutorial](https://docs.llamaindex.ai/en/stable/examples/agent/react_agent/) as the starting point)
+<!-- (we will use [this example](https://docs.llamaindex.ai/en/stable/examples/agent/agent_runner/agent_runner/) as a starting point, since it reimplements the function calling agent and returns a step-by step breakdown of its thought process). -->
+4. Have a look at the prompt, the agent gives to the LLM (you can find it using `agent.get_prompts()`)
+5. Discuss the prompt with the group! What does it do? How does it do it?
+6. Ask the agent to tell you the weather in a given location.
+7. Watch in LM Studio how the LLM called by the agent creates the thought process, function calls and the final response.
+8. Try to break the agent by asking stuff it cannot answer. Be creative. (On one occasion I just said "Hi" and it went into infinity loop because it did not need a tool for that and there wasn't a "none" tool 😜.)
```

Comment:
Funny anecdote, I like it here

But shouldn't it read:
..it went into an infinite loop